### ISSUE TITLE
Implementing a siamese loss function

### ISSUE THREAD

**Issue by vbalnt** [twitter/torch-autograd#94]
Hi I have been trying to implement a Siamese kind of loss function, but I cant figure out how to compute the embeddings for two different inputs inside the autograd function

Here is what I am trying to do 

```lua
t = require 'torch'
grad = require 'autograd'
require 'cutorch'
require 'cunn'
require 'cudnn'

local model = nn.Sequential()
model:add(cudnn.SpatialConvolution(1, 16, 3, 3))
model:add(cudnn.Tanh())
model:add(nn.Reshape(16*30*30))
model:add(nn.Linear(16*30*30, 128))
model:add(cudnn.Tanh())
model:cuda()

local modelf, params = grad.functionalize(model)

neuralNet = function(params, x)
   local x1 = x[1]
   local x2 = x[2]

   local h = modelf(params, x1)
   local h2 = modelf(params, x2)
   return torch.sum(torch.abs(h-h2))
end

dneuralNet = grad(neuralNet)

tt = torch.rand(1,1,32,32):cuda()
tt2 = torch.rand(1,1,32,32):cuda()

dparams, loss = dneuralNet(params, {tt,tt2})
print(loss)
```

This does not work since `h==h2`, so loss comes out 0

Is there any way to do something similar?

**Comment by alexbw** [comment]
That's pretty non-obvious behavior, unfortunately, and we should probably fix it.

This works for me:

```lua
t = require 'torch'
grad = require 'autograd'

local model = nn.Sequential()
model:add(nn.SpatialConvolution(1, 16, 3, 3))
model:add(nn.Tanh())
model:add(nn.Reshape(16*30*30))
model:add(nn.Linear(16*30*30, 128))
model:add(nn.Tanh())


local modelf, params = grad.functionalize(model)

neuralNet = function(params, x)
   local x1 = x[1]
   local x2 = x[2]

   local h = torch.clone(modelf(params, x1))
   local h2 = torch.clone(modelf(params, x2))
   return torch.sum(torch.abs(h-h2))
end

dneuralNet = grad(neuralNet)

tt = torch.rand(1,1,32,32)
tt2 = torch.rand(1,1,32,32)

dparams, loss = dneuralNet(params, {tt,tt2})
print(loss)
```

**Comment by vbalnt** [comment]
Now that I look at it, it does seem logical, this is also how the torch `nn` models work. However, maybe a mention in the docs could be useful. Thanks & closing.
